### PR TITLE
accounts-db/read-cache: don't populate read cache for withdraw, deposit, fee_distribution and store_capital update

### PR DESCRIPTION
**Pull request by HaoranYi**
#### Problem

When performing withdraw, deposit, fee_distribution, and store_capital_update, we do load, update and store. We put the old account into read cache after loading. This is not needed and can lead to inefficient use of read cache. 

#### Summary of Changes

- expose `load_with` API from accounts-db to bank
- create a helper method `get_account_with_fixed_root_no_cache`
- use `get_account_with_fixed_root_no_cache` to avoid putting stale account
  into read cache.

#### Performance comparison

image

blue: base 
red: this PR

It shows better `load_us` and `prog_cache_us`. `prog_cache_us` includes filtering which load accounts and checks owners. Both of this operations benefits from more efficient use of read-only caches.  

Fixes #

### CODE REVIEW

**Comment by HaoranYi** [comment]
A result from a longer run (12h)

image

image

File: runtime/src/bank.rs
Comment (brooksprumo) [inline runtime/src/bank.rs]: Can the lifetimes in the `impl for` be `'local`? This is just personal preference, as I find the `impl for` syntax confusing how you can reuse/duplicate lifetime names.

```suggestion
        callback: impl for<'local> Fn(&'local AccountSharedData) -> bool,
```

**Review by brooksprumo — COMMENTED** [review]
It looks like if there's a scenario where the same account is getting loaded by multiple things in parallel, that *not* putting the account into the read cache could cause a different load to hit the disk again. 

I don't know how likely this is to happen though. Both through how accounts are write-locked by transactions, and all the functions modified by this PR call `store_account()` after calling `get_account_with_fixed_root_no_cache()`, so the time window between these two is pretty small (yes, the thread could get switched out by the OS).

I'm inclined to think this pathological case isn't an issue, but did want to bring it up to get thoughts from others.

**Comment by HaoranYi** [comment]
> It looks like if there's a scenario where the same account is getting loaded by multiple things in parallel, that _not_ putting the account into the read cache could cause a different load to hit the disk again.
> 
> I don't know how likely this is to happen though. Both through how accounts are write-locked by transactions, and all the functions modified by this PR call `store_account()` after calling `get_account_with_fixed_root_no_cache()`, so the time window between these two is pretty small (yes, the thread could get switched out by the OS).
> 
> I'm inclined to think this pathological case isn't an issue, but did want to bring it up to get thoughts from others.

I don't think that there could be parallel caller for these functions concurrently within the same slot. If this happens, we will have a race condition.
One case that's helpful to put into the cache is during fork. If both forks need to access the same account, putting into the cache will be helpful.

**Comment by Lichtso** [comment]
> A result from a longer run (12h)

> blue: base, red: this PR

Same color scheme?

**Review by Lichtso — APPROVED** [review]

**Comment by HaoranYi** [comment]
blue: base
red: this PR

**Review by brooksprumo — APPROVED** [review]
Looks good to me. I'd like to get @jeffwashington 's approval too before merging

**Review by jeffwashington — APPROVED** [review]
looks like a great idea. not sure how often these things would be loaded from append vecs vs write cache initially. This certainly can't harm anything.

**Pull request merged**
